File: frontend/src/services/crm.js
import { getToken, getBusinessId } from './auth';

const BASE = '/api/crm';

function headers() {
  const h = { 'Content-Type': 'application/json' };
  const t = getToken();
  if (t) h['Authorization'] = `Bearer ${t}`;
  const b = getBusinessId();
  if (b) h['X-Business-Id'] = b;
  return h;
}

async function req(path, opts = {}) {
  const res = await fetch(`${BASE}${path}`, { ...opts, headers: { ...headers(), ...(opts.headers || {}) } });
  if (res.status === 401) { window.location.href = '/login'; throw new Error('Session expired'); }
  if (!res.ok) {
    const txt = await res.text();
    let msg = txt;
    try { msg = JSON.parse(txt).detail || txt; } catch {}
    throw new Error(msg || `HTTP ${res.status}`);
  }
  return res.json();
}

// Overview / pipeline
export const crmOverview = () => req('/overview');
export const pipeline = () => req('/pipeline');

// Companies
export const listCompanies = (search = '') =>
  req(`/companies${search ? `?search=${encodeURIComponent(search)}` : ''}`);
export const getCompany = (id) => req(`/companies/${id}`);
export const createCompany = (body) => req('/companies', { method: 'POST', body: JSON.stringify(body) });
export const updateCompany = (id, body) => req(`/companies/${id}`, { method: 'PATCH', body: JSON.stringify(body) });
export const deleteCompany = (id) => req(`/companies/${id}`, { method: 'DELETE' });

// Contacts
export const listContacts = (opts = {}) => {
  const qs = new URLSearchParams();
  if (opts.search) qs.set('search', opts.search);
  if (opts.company_id) qs.set('company_id', opts.company_id);
  const s = qs.toString();
  return req(`/contacts${s ? '?' + s : ''}`);
};
export const getContact = (id) => req(`/contacts/${id}`);
export const createContact = (body) => req('/contacts', { method: 'POST', body: JSON.stringify(body) });
export const updateContact = (id, body) => req(`/contacts/${id}`, { method: 'PATCH', body: JSON.stringify(body) });
export const deleteContact = (id) => req(`/contacts/${id}`, { method: 'DELETE' });

// Deals
export const listDeals = (opts = {}) => {
  const qs = new URLSearchParams();
  if (opts.stage) qs.set('stage', opts.stage);
  if (opts.search) qs.set('search', opts.search);
  const s = qs.toString();
  return req(`/deals${s ? '?' + s : ''}`);
};
export const getDeal = (id) => req(`/deals/${id}`);
export const createDeal = (body) => req('/deals', { method: 'POST', body: JSON.stringify(body) });
export const updateDeal = (id, body) => req(`/deals/${id}`, { method: 'PATCH', body: JSON.stringify(body) });
export const deleteDeal = (id) => req(`/deals/${id}`, { method: 'DELETE' });

// Interactions
export const listInteractions = (opts = {}) => {
  const qs = new URLSearchParams();
  for (const [k, v] of Object.entries(opts)) if (v) qs.set(k, v);
  const s = qs.toString();
  return req(`/interactions${s ? '?' + s : ''}`);
};
export const createInteraction = (body) => req('/interactions', { method: 'POST', body: JSON.stringify(body) });
export const deleteInteraction = (id) => req(`/interactions/${id}`, { method: 'DELETE' });

// AI-drafted outreach — returns { variants: [{ tone, subject, body }, ...] }.
// Privacy: this endpoint runs the prompt with sensitive=True on the backend
// so the LLM call stays on local Ollama.
export const draftOutreach = (contactId) =>
  req(`/contacts/${contactId}/draft-outreach`, { method: 'POST' });

// AI lead scoring — returns { score, bucket, reason, scored_at, icp_set }.
// score is 0-100, bucket is "high"/"medium"/"low"/"spam"/null, both null
// when no ICP is set yet (the response carries icp_set=false in that case).
export const scoreContactFit = (contactId) =>
  req(`/contacts/${contactId}/score-fit`, { method: 'POST' });

// Workspace ICP description — used by the scorer. The CRM `req` helper
// already prefixes /api/crm; for the workspace path we call fetch directly
// using the same auth headers.
const WS_BASE = '/api/workspace';
function _wsHeaders() {
  const h = { 'Content-Type': 'application/json' };
  const t = getToken(); if (t) h.Authorization = `Bearer ${t}`;
  const b = getBusinessId(); if (b) h['X-Business-Id'] = b;
  return h;
}
export const readIcp = async () => {
  const r = await fetch(`${WS_BASE}/icp`, { headers: _wsHeaders() });
  if (!r.ok) throw new Error(await r.text());
  return r.json();
};
export const writeIcp = async (icp_description) => {
  const r = await fetch(`${WS_BASE}/icp`, {
    method: 'PUT', headers: _wsHeaders(),
    body: JSON.stringify({ icp_description }),
  });
  if (!r.ok) throw new Error(await r.text());
  return r.json();
};

// Email-paste lead capture: extract preview, then save with edits.
export const extractEmail = (raw_email, override_email = null) =>
  req('/leads/extract-email', {
    method: 'POST',
    body: JSON.stringify({ raw_email, override_email }),
  });
export const saveLeadFromEmail = (payload) =>
  req('/leads/from-email', {
    method: 'POST',
    body: JSON.stringify(payload),
  });

// BANT extraction from a pasted prospect reply. Returns the structured
// signals + an optional `suggested_stage` the user can advance the deal
// to with one click. Privacy: prompt sees reply content → sensitive=True.
export const extractBant = (contactId, reply_text) =>
  req(`/contacts/${contactId}/extract-bant`, {
    method: 'POST',
    body: JSON.stringify({ reply_text }),
  });

export const DEAL_STAGES = ['lead', 'qualified', 'proposal', 'negotiation', 'won', 'lost'];
export const INTERACTION_TYPES = ['call', 'email', 'meeting', 'note'];

File: api/routers/crm.py
"""CRM router — companies, contacts, deals, interactions, pipeline, bulk ops."""
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends

from api import crm as _crm
from api.auth import get_current_context

router = APIRouter(tags=["crm"])


@router.get("/api/crm/overview")
def crm_overview(ctx: dict = Depends(get_current_context)):
    return _crm.crm_overview(ctx["business_id"])


@router.get("/api/crm/pipeline")
def pipeline_api(ctx: dict = Depends(get_current_context)):
    return _crm.deal_pipeline_stats(ctx["business_id"])


# ── Companies ────────────────────────────────────────────────────────────────
@router.get("/api/crm/companies")
def list_companies_api(search: Optional[str] = None, limit: int = 100,
                       ctx: dict = Depends(get_current_context)):
    return _crm.list_companies(ctx["business_id"], search=search, limit=limit)


@router.post("/api/crm/companies")
def create_company_api(body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.create_company(ctx["business_id"], ctx["user"]["id"], body)


@router.get("/api/crm/companies/{company_id}")
def get_company_api(company_id: str, ctx: dict = Depends(get_current_context)):
    return _crm.get_company(ctx["business_id"], company_id)


@router.patch("/api/crm/companies/{company_id}")
def update_company_api(company_id: str, body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.update_company(ctx["business_id"], company_id, body)


@router.delete("/api/crm/companies/{company_id}")
def delete_company_api(company_id: str, ctx: dict = Depends(get_current_context)):
    _crm.delete_company(ctx["business_id"], company_id)
    return {"ok": True}


@router.post("/api/crm/companies/bulk-delete")
def bulk_delete_companies_api(body: dict, ctx: dict = Depends(get_current_context)):
    ids = body.get("ids") or []
    return {"deleted": _crm.bulk_delete_companies(ctx["business_id"], ids)}


# ── Contacts ─────────────────────────────────────────────────────────────────
@router.get("/api/crm/contacts")
def list_contacts_api(search: Optional[str] = None, company_id: Optional[str] = None,
                      limit: int = 100, ctx: dict = Depends(get_current_context)):
    return _crm.list_contacts(ctx["business_id"], search=search, company_id=company_id, limit=limit)


@router.post("/api/crm/contacts")
def create_contact_api(body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.create_contact(ctx["business_id"], ctx["user"]["id"], body)


@router.get("/api/crm/contacts/{contact_id}")
def get_contact_api(contact_id: str, ctx: dict = Depends(get_current_context)):
    return _crm.get_contact(ctx["business_id"], contact_id)


@router.patch("/api/crm/contacts/{contact_id}")
def update_contact_api(contact_id: str, body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.update_contact(ctx["business_id"], contact_id, body)


@router.delete("/api/crm/contacts/{contact_id}")
def delete_contact_api(contact_id: str, ctx: dict = Depends(get_current_context)):
    _crm.delete_contact(ctx["business_id"], contact_id)
    return {"ok": True}


@router.post("/api/crm/contacts/bulk-delete")
def bulk_delete_contacts_api(body: dict, ctx: dict = Depends(get_current_context)):
    ids = body.get("ids") or []
    return {"deleted": _crm.bulk_delete_contacts(ctx["business_id"], ids)}


# ── Deals ────────────────────────────────────────────────────────────────────
@router.get("/api/crm/deals")
def list_deals_api(stage: Optional[str] = None, search: Optional[str] = None,
                   limit: int = 200, ctx: dict = Depends(get_current_context)):
    return _crm.list_deals(ctx["business_id"], stage=stage, search=search, limit=limit)


@router.post("/api/crm/deals")
def create_deal_api(body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.create_deal(ctx["business_id"], ctx["user"]["id"], body)


@router.get("/api/crm/deals/{deal_id}")
def get_deal_api(deal_id: str, ctx: dict = Depends(get_current_context)):
    return _crm.get_deal(ctx["business_id"], deal_id)


@router.patch("/api/crm/deals/{deal_id}")
def update_deal_api(deal_id: str, body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.update_deal(ctx["business_id"], deal_id, body)


@router.delete("/api/crm/deals/{deal_id}")
def delete_deal_api(deal_id: str, ctx: dict = Depends(get_current_context)):
    _crm.delete_deal(ctx["business_id"], deal_id)
    return {"ok": True}


@router.post("/api/crm/deals/bulk-delete")
def bulk_delete_deals_api(body: dict, ctx: dict = Depends(get_current_context)):
    ids = body.get("ids") or []
    return {"deleted": _crm.bulk_delete_deals(ctx["business_id"], ids)}


@router.post("/api/crm/deals/bulk-stage")
def bulk_stage_deals_api(body: dict, ctx: dict = Depends(get_current_context)):
    ids = body.get("ids") or []
    stage = body.get("stage") or ""
    return {"updated": _crm.bulk_update_deal_stage(ctx["business_id"], ids, stage)}


# ── Interactions ─────────────────────────────────────────────────────────────
@router.get("/api/crm/interactions")
def list_interactions_api(contact_id: Optional[str] = None, company_id: Optional[str] = None,
                          deal_id: Optional[str] = None, limit: int = 100,
                          ctx: dict = Depends(get_current_context)):
    return _crm.list_interactions(
        ctx["business_id"],
        contact_id=contact_id, company_id=company_id, deal_id=deal_id, limit=limit,
    )


@router.post("/api/crm/interactions")
def create_interaction_api(body: dict, ctx: dict = Depends(get_current_context)):
    return _crm.create_interaction(ctx["business_id"], ctx["user"]["id"], body)


@router.delete("/api/crm/interactions/{interaction_id}")
def delete_interaction_api(interaction_id: str, ctx: dict = Depends(get_current_context)):
    _crm.delete_interaction(ctx["business_id"], interaction_id)
    return {"ok": True}
